servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -469,18 +469,23 @@ public long getWorkerId(NetAddress workerNetAddress) {
    * Updates metadata when a worker registers with the master. Called by workers via RPC.
    *
    * @param workerId the worker id of the worker registering
+   * @param workerAddress the worker NetAddress of the worker registering
    * @param totalBytesOnTiers list of total bytes on each tier
    * @param usedBytesOnTiers list of the used byes on each tier
    * @param currentBlocksOnTiers a mapping of each storage dir, to all the blocks on that storage
    * @return the worker id
    */
-  public long workerRegister(long workerId, List<Long> totalBytesOnTiers,
+  public long workerRegister(long workerId, NetAddress workerAddress, List<Long> totalBytesOnTiers,
       List<Long> usedBytesOnTiers, Map<Long, List<Long>> currentBlocksOnTiers) {
     synchronized (mBlocks) {
       synchronized (mWorkers) {
         if (!mWorkers.contains(mIdIndex, workerId)) {
```

Comment:
In master fault tolerant mode, when a standby master is elected as the new leader, the worker will send heartbeat to the new leader, but since the new leader knows no previous worker(no related information is journaled), the leader will send back a heartbeat with `Register` command, so the worker will finally call `workerRegister` here. 

In previous implementation, since `mWorkers` is empty, this block will be executed and `-1` will be returned to worker, the worker will think that `-1` is the new worker ID assigned by the new master, so it changes its previous ID to `-1`, and keep trying to send heartbeat to the new master with ID `-1`, but the new master won't understand ID `-1`, so it lets the worker register again, then an infinite loop happens.